File: Takonaut/Sources/TNMazeGenerator.m
#import "TNMazeGenerator.h"

typedef NS_ENUM(NSUInteger, DirectionType) {
  DTNorth,
  DTSouth,
  DTEast,
  DTWest
};

@interface TNMazeGenerator()
@property(nonatomic,assign) NSUInteger width;
@property(nonatomic,assign) NSUInteger height;
@property(nonatomic,assign) CGPoint start;
@end

@implementation TNMazeGenerator

- (MazeTyleType **)calculateMaze:(NSUInteger)row col:(NSUInteger)col startingPosition:(CGPoint)position
{
  _width	= col;
  _height	= row;
  _start = position;
  
  //--- init maze ---//
  MazeTyleType **maze = (MazeTyleType **)calloc(_height, sizeof(MazeTyleType *));
  for (int r = 0; r < _height;++r)
  {
   maze[r] = (MazeTyleType *)calloc(_width, sizeof(MazeTyleType));
  }

  int posX = (int)self.start.x;
  int posY = (int)self.start.y;

  //--- taking start tile ---//
  maze[posX][posY] = MTStart;

  NSMutableArray<NSDictionary *> *visitedTiles = [@[] mutableCopy];
  NSMutableArray *currentPath = [@[@{@"x" : @(posX), @"y" : @(posY)}] mutableCopy];
  while ([currentPath count])
  {
    //--- digging in some diretions ---//
    NSMutableArray<NSNumber *> *possibleDirections = [NSMutableArray<NSNumber *> array];
    if ((posX - 2 >= 0) && (maze[posX - 2][posY] == MTWall))
    {
      [possibleDirections addObject:@(DTNorth)];
    }
    
    if ((posX + 2 < self.height) && (maze[posX + 2][posY] == MTWall))
    {
      [possibleDirections addObject:@(DTSouth)];
    }
    
    if ((posY + 2 < self.width) && (maze[posX][posY + 2] == MTWall))
    {
      [possibleDirections addObject:@(DTEast)];
    }
    
    if ((posY - 2 >= 0) && (maze[posX][posY - 2] == MTWall))
    {
      [possibleDirections addObject:@(DTWest)];
    }
    
    if (possibleDirections.count > 0) // forward
    {
      DirectionType direction = [possibleDirections[arc4random() % possibleDirections.count] shortValue];
      switch (direction)
      {
        case DTNorth: {
          maze[posX - 2][posY] = MTPath;
          maze[posX - 1][posY] = MTPath;
          posX -= 2;
          break;
        }
        case DTSouth: {
          maze[posX + 2][posY] = MTPath;
          maze[posX + 1][posY] = MTPath;
          posX += 2;
          break;
        }
        case DTEast: {
          maze[posX][posY + 2] = MTPath;
          maze[posX][posY + 1] = MTPath;
          posY += 2;
          break;
        }
        case DTWest: {
          maze[posX][posY - 2] = MTPath;
          maze[posX][posY - 1] = MTPath;
          posY -= 2;
          break;
        }
      }
      [currentPath addObject:@{@"x" : @(posX), @"y" : @(posY)}];
      [visitedTiles addObject:@{@"x" : @(posX), @"y" : @(posY), @"steps" : @(currentPath.count * 2)}];
    }
    else //backtracking
    {
      NSDictionary *back = [currentPath lastObject];
      posX = [back[@"x"] intValue];
      posY = [back[@"y"] intValue];
      [currentPath removeLastObject];
    }
  }
  
  //--- taking end tile ---//
  NSDictionary *end;
  for (NSDictionary *tile in visitedTiles)
  {
    if ([tile[@"steps"] intValue] >= [end[@"steps"] intValue])
    {
      end = tile;
    }
  }
  maze[[end[@"x"] intValue]][[end[@"y"] intValue]] = MTEnd;
  return maze;
}

@end

File: Takonaut/Sources/TNMazeGenerator.cpp
//
//  TNMazeGenerator.cpp
//  Takonaut
//
//  Created by mugx on 27/04/16.
//  Copyright © 2016 mugx. All rights reserved.
//

#include "TNMazeGenerator.hpp"
#include <stdlib.h>
#include <iostream>
#include <list>
#include <vector>
#include <map>

struct Tile {
  unsigned int x;
  unsigned int y;
  unsigned long steps;
  
  Tile()
  {
    x = 0;
    y = 0;
    steps = 0;
  }
};

MazeTyleType** TNMazeGenerator::calculateMaze(int startX, int startY, int width, int height)
{
  //--- init maze ---//
  MazeTyleType **maze = (MazeTyleType **)calloc(height, sizeof(MazeTyleType *));
  for (int r = 0; r < height;++r)
  {
    maze[r] = (MazeTyleType *)calloc(width, sizeof(MazeTyleType));
  }
  
  //--- taking start tile ---//
  int posX = startX;
  int posY = startY;
  maze[startX][startY] = MTStart;
  Tile initialTile;
  initialTile.x = startX;
  initialTile.y = startY;
  std::list<Tile> visitedTiles;
  std::list<Tile> currentPath = {initialTile};
  
  while (currentPath.size())
  {
    std::vector<char> possibleDirections;
    if ((posX - 2 >= 0) && (maze[posX - 2][posY] == MTWall))
    {
      possibleDirections.push_back('n');
    }
    
    if ((posX + 2 < height) && (maze[posX + 2][posY] == MTWall))
    {
      possibleDirections.push_back('s');
    }
    
    if ((posY + 2 < width) && (maze[posX][posY + 2] == MTWall))
    {
      possibleDirections.push_back('e');
    }
    
    if ((posY - 2 >= 0) && (maze[posX][posY - 2] == MTWall))
    {
      possibleDirections.push_back('w');
    }
    
    if (possibleDirections.size() > 0) // forward
    {
      char direMTion = possibleDirections[arc4random() % possibleDirections.size()];
      switch (direMTion)
      {
        case 'n': {
          maze[posX - 2][posY] = MTPath;
          maze[posX - 1][posY] = MTPath;
          posX -= 2;
          break;
        }
        case 's': {
          maze[posX + 2][posY] = MTPath;
          maze[posX + 1][posY] = MTPath;
          posX += 2;
          break;
        }
        case 'e': {
          maze[posX][posY + 2] = MTPath;
          maze[posX][posY + 1] = MTPath;
          posY += 2;
          break;
        }
        case 'w': {
          maze[posX][posY - 2] = MTPath;
          maze[posX][posY - 1] = MTPath;
          posY -= 2;
          break;
        }
      }
      Tile tile;
      tile.x = posX;
      tile.y = posY;
      tile.steps = currentPath.size() * 2;
      currentPath.push_back(tile);
      visitedTiles.push_back(tile);
    }
    else //backtracking
    {
      Tile back = currentPath.back();
      posX = back.x;
      posY = back.y;
      currentPath.pop_back();
    }
  }
  
  //--- taking end tile ---//
  Tile end;
  for (Tile tile : visitedTiles)
  {
    if (tile.steps >= end.steps)
    {
      end = tile;
    }
  }
  maze[end.x][end.y] = MTEnd;
  return maze;
}